### PR TITLE
Schedule tasks which are futures to completion

### PR DESCRIPTION
**Pull request by jiahaog**
Today, `scheduleTask` to does not execute the scheduled task to completion, when the task is a future. If we are providing a `Future<int> Function()` to `scheduleTask`, the current implementation of `scheduleTask` returns a future when this task *starts*, but not when it completes. To actually run the task to completion, the callsite needs to look something like the following, which is kind of awkward:

```dart
void foo() async {
  Future<int> Function() task = () async => 1;

  // Awkward?
  await await binding.scheduleTask(task, ...);
}
```

I stumbled across this when using the API above but with a single `await`, and I was quite confused why the `unawaited_futures` lint was triggering which led me to looking deeper at the issue.

It would be desirable for task to be scheduled to completion, even when `task` is a function that returns a future for a simpler API. We can already opt-into this behavior "for free" when the future is [handed to completer](https://github.com/flutter/flutter/blob/9f5174f7210c4af54e681927aacea7da14cc0fe9/packages/flutter/lib/src/scheduler/binding.dart#L87). The signature of [`Completer.complete`](https://api.flutter.dev/flutter/dart-async/Completer/complete.html) is:

```dart
void complete(
  [FutureOr<T>? value]
)
```

From the docs (emphasis mine):

> **[case 1]** Completes future with the supplied values.
> ...
> **[case 2]** If the value is itself a future, the completer will wait for that future to complete, and complete with the same result, whether it is a success or an error.

Today, we're hitting the first case, but with the correct type annotations in this PR, we could get the second.

I also tested this change in an internal presubmit (tap/476354531), and it passes tests so it doesn't look like a breaking change.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jiahaog** [comment]
Friendly ping - @chunhtai can you please take a look :slightly_smiling_face:

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 1]
```diff
@@ -50,7 +50,7 @@ typedef FrameCallback = void Function(Duration timeStamp);
 ///
 /// The type argument `T` is the task's return value. Consider `void` if the
 /// task does not return a value.
-typedef TaskCallback<T> = T Function();
+typedef TaskCallback<T> = FutureOr<T> Function();
```
Comment (chunhtai) [inline packages/flutter/lib/src/scheduler/binding.dart:53]: We should add some documentation in scheduleTask method to mention the behavior.

**Review by chunhtai — APPROVED** [review]
LGTM, left a documentation nit

File: packages/flutter/lib/src/scheduler/binding.dart (hunk 1 above)
Comment (jiahaog) [reply packages/flutter/lib/src/scheduler/binding.dart:53]: Thanks! Looking at `scheduleTask`, the doc comment mentions:

> Schedules the given task with the given priority and returns a [Future](https://api.flutter.dev/flutter/dart-async/Future-class.html) that completes to the task's eventual return value.

This seems to already imply that the task will complete (irregardless of whether the task is async or not). In that case do you think we should still need to update it?

Comment (chunhtai) [reply packages/flutter/lib/src/scheduler/binding.dart:53]: `eventual return value` part is a bit vague. maybe add phrase to describe the behavior more clearly.

Comment (jiahaog) [reply packages/flutter/lib/src/scheduler/binding.dart:53]: Done, would you like to take a look again?

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**